### ISSUE TITLE
Autoscaling on log axis with negative data results in arbitrary limits.

### ISSUE THREAD

**Issue by ImportanceOfBeingErnest** [matplotlib/matplotlib#10782]
### Bug report

**Bug summary**

Plotting negative values on a log scale is possible. However it changes the axis limits in a seemingly arbitrary way. 

**Code for reproduction**

The following shows three scatter plots. The `x` values
(1) are all positive
(2) contain one negative value
(3) have the negative value filtered out

```
import numpy as np; np.random.seed(1)
import matplotlib.pyplot as plt

x = np.random.rayleigh(100,400)
y = np.random.rand(len(x))

fig, (ax,ax2, ax3) = plt.subplots(nrows=3)

ax.scatter(x,y)
ax.set_xscale("log")
ax.set_title("Scatter of all-positive values")

ax2.scatter(x-2,y)
ax2.set_xscale("log")
ax2.set_title("Scatter with one negative value")

cond = x-2 > 0
ax3.scatter((x-2)[cond],y[cond])
ax3.set_xscale("log")
ax3.set_title("Expected scatter with one negative value")

plt.tight_layout()
plt.show()
```

**Actual outcome**

The second subplot, where the data contains a negative value, shows only part of the valid data, because the axis is autoscaled to some seemingly arbitrary region.

![image](https://user-images.githubusercontent.com/23121882/37407656-7a2ba57c-279a-11e8-8957-1f85be04d97e.png)

**Expected outcome**

The third subplot would be the expected outcome with the axis autoscaled to show all points. 

One could also argue that plotting negative data on a log scale should not be supported at all,  and that it should be the user's responsibility to make sure the data is positive. In that case however it would make sense to issue a warning instead of just plotting something arbitrary.

**Matplotlib version**

  * Operating system: Windows 8.1
  * Matplotlib version: 2.2
  * Matplotlib backend: any
  * Python version: 2.7.10

**Comment by jklymak** [comment]
What happens if you set the scale before calling scatter?  Loglog etc check for negative I think

**Comment by ImportanceOfBeingErnest** [comment]
Good point, but setting the scale before the plotting does not make any difference, i.e.

```
ax2.set_xscale("log")
ax2.scatter(x-2,y)
```

results in the same plot as above.

**Comment by dstansby** [comment]
I think this might be a duplicate of https://github.com/matplotlib/matplotlib/issues/7733

**Comment by ImportanceOfBeingErnest** [comment]
@dstansby Not sure; the other post seems to give the correct result (0 not being in the plot, but the axis limits are correct and the plot shows all points.). Setting bottom=0 may not make much sense for a log axis, so the warning is expected? 

Here, we have no warning and limits are just arbitrarily shifted.

**Comment by dstansby** [comment]
I still slightly suspect that they might have a common cause, but I'm not sure. I'll try and dig into this one to see what's happening.

**Comment by QuLogic** [comment]
This was fixed by #18642.
